plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -3630,6 +3634,38 @@ public void testEmptyDelete()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testCtasFromAVersionedTable()
+            throws Exception
+    {
+        String sourceTableName = "test_ctas_versioned_source_" + randomTableSuffix();
+        String sinkTableNameVersion = "test_ctas_versioned_sink_" + randomTableSuffix();
+        String sinkTableNameTimestamp = "test_ctas_versioned_sink_" + randomTableSuffix();
+
+        assertUpdate("CREATE TABLE " + sourceTableName + "(an_integer integer)");
+        Thread.sleep(1);
+        assertUpdate("INSERT INTO " + sourceTableName + " VALUES (1), (2), (3)", 3);
+        long afterInsert123SnapshotId = getLatestSnapshotId(sourceTableName);
+        long afterInsert123EpochMillis = getCommittedAtInEpochMilliSeconds(sourceTableName, afterInsert123SnapshotId);
+        Thread.sleep(1);
+        assertUpdate("INSERT INTO " + sourceTableName + " VALUES (4), (5), (6)", 3);
+        long afterInsert456SnapshotId = getLatestSnapshotId(sourceTableName);
+        Thread.sleep(1);
+        assertUpdate("DELETE FROM " + sourceTableName + " WHERE an_integer < 5", 4);
+
+        // Clone the source table at the state before the deletion occurred to a sink table
+        assertUpdate("CREATE TABLE " + sinkTableNameVersion + " AS SELECT * FROM " + sourceTableName + " FOR VERSION AS OF " + afterInsert456SnapshotId, 6);
+        assertUpdate("CREATE TABLE " + sinkTableNameTimestamp + " AS SELECT * FROM " + sourceTableName + " FOR TIMESTAMP AS OF " + timestampLiteral(afterInsert123EpochMillis, 9), 3);
```

Comment:
Here seems to be a problem/bug in Iceberg.

`table.history()` in the method seems to have only `1` entry, even though it should have had `4`

On debug it seems to me that there are `5` snapshots (`1` too many) at the time of executing this CTAS statement.

https://github.com/trinodb/trino/blob/c19dfc1ad19322d5583c7dbda63830f873cb101d/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L477-L485

Reply:
Side discovery while looking on the current problem:

The `DELETE` statement is creating two table snapshots:

https://github.com/trinodb/trino/blob/2a466edc37e26047d8cbc7f98023aa62452e782c/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1569


https://github.com/trinodb/trino/blob/2a466edc37e26047d8cbc7f98023aa62452e782c/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1598


I think that this is not really intended.

cc @alexjo2144 @findepi

Reply:
Created  bug https://github.com/trinodb/trino/issues/12843

Reply:
We can just rework the test to use inserts while I'm looking at that, right?